CONTRIBUTING.md

```diff
@@ -111,73 +143,80 @@ Make sure you document each commit and squash them appropriately. You can use th
+If you're working on a bug fix, it's a good idea to start with the `1.$MINOR.x` branch, since we can always safely merge from stable to `1.x`, but not other way around.
 
-4. To build the launcher and publish all components locally,
+### Instruction to build all modules from source
 
-		$ sbt
-		> publishLocal
+1. Install the current stable binary release of sbt (see [Setup]), which will be used to build sbt from source.
+2. Get the source code.
 
-5. To use this locally built version of sbt, copy your stable `~/bin/sbt` script to `~/bin/xsbt` and change it to use the launcher jar at `<sbt>/launch/target/sbt-launch.jar`.
+   ```
+   $ mkdir sbt-modules
+   $ cd sbt-modules
+   $ for i in sbt io util librarymanagement zinc; do \
+     git clone https://github.com/sbt/$i.git && (cd $i; git checkout -b 1.1.x origin/1.1.x)
+   done
+   $ cd sbt
+   $ ./sbt-allsources.sh
+   ```
 
-	Directory `target` is removed by `clean` command. Second solution is using the artifact stored in the local ivy repository.
+3. To build and publish all components locally,
 
-	The launcher is located in:
+   ```
+   $ ./sbt-allsources.sh
+   sbt:sbtRoot> publishLocalAllModule
+   ```
 
-    $HOME/.ivy2/local/org.scala-sbt/sbt-launch/0.13.9/jars/sbt-launch.jar
+### Instruction to build just sbt
 
-	for v0.13.9 tag, or in:
+If the change you are making is contained in sbt/sbt, you could publishLocal on sbt/sbt:
 
-    $HOME/.ivy2/local/org.scala-sbt/sbt-launch/0.13.10-SNAPSHOT/jars/sbt-launch.jar
+```
+$ sbt
+sbt:sbtRoot> publishLocal
+```
 
-	for the development branch.
+### Using the locally built sbt
 
-## Modifying sbt
+To use the locally built sbt, set the version in `build.properties` file to `1.$MINOR.$PATCH-SNAPSHOT`.
 
-1. When developing sbt itself, run `compile` when checking compilation only.
+```
+$ cd ../hello
+$ sbt
+> compile
+```
 
-2. To use your modified version of sbt in a project locally, run `publishLocal`.
+### Clearing out boot and local cache
 
-3. After each `publishLocal`, clean the `~/.sbt/boot/` directory.  Alternatively, if sbt is running and the launcher hasn't changed, run `reboot full` to have sbt do this for you.
+When you run a locally built sbt, the JAR artifacts will be now cached under `$HOME/.sbt/boot/scala-2.12.4/org.scala-sbt/sbt/1.$MINOR.$PATCH-SNAPSHOT` directory. To clear this out run: `reboot dev` command from sbt's session of your test application.
 
-4. If a project has `project/build.properties` defined, either delete the file or change `sbt.version` to `1.0.0-SNAPSHOT`.
+One drawback of `-SNAPSHOT` version is that it's slow to resolve as it tries to hit all the resolvers. You can workaround that by using a version name like `1.$MINOR.$PATCH-LOCAL1`. A non-SNAPSHOT artifacts will now be cached under `$HOME/.ivy/cache/` directory, so you need to clear that out using [sbt-dirty-money](https://github.com/sbt/sbt-dirty-money)'s `cleanLocal` task.
 
```

Comment:
Do you mean ``$HOME/.ivy/local/` here?

Reply:
no I think the mistake is the suggestion should be to run `cleanCache`

Reply:
I've just been incrementing my version number to reduce the cognitive load :)